### PR TITLE
Remove `groupedInstanceKeys` property from `parentKeys`

### PR DESCRIPTION
**Pull request by JonasDov**
We were redundantly storing all `groupedInstanceKeys` both in the `key` of the node, and in the `parentKeys`. Removed it from `parentKeys`.
Also, we had an issue where with a lot of items we would get out of memory exception, so changed to not use array deconstruction.
Finally, removed stringification of `parentKeys` as it was taking a lot of time.

### CODE REVIEW

**Review by grigasp — CHANGES_REQUESTED** [review]

File: packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts [hunk 1]
```diff
@@ -141,47 +141,59 @@ export type StandardHierarchyNodeKey = InstancesNodeKey | GroupingNodeKey;
...
+ * A key that uniquely identifies parent grouping node in a hierarchy level.
+ * This key differs from [[GroupingNodeKey]] by not having `groupedInstanceKeys` property.
+ * @beta
+ */
+export type ParentGroupingHierarchyNodeKey =
+  | Omit<ClassGroupingNodeKey, "groupedInstanceKeys">
+  | Omit<LabelGroupingNodeKey, "groupedInstanceKeys">
+  | Omit<PropertyValueRangeGroupingNodeKey, "groupedInstanceKeys">
+  | Omit<PropertyOtherValuesGroupingNodeKey, "groupedInstanceKeys">
+  | Omit<PropertyValueGroupingNodeKey, "groupedInstanceKeys">;
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts]: Omit does not work with union types. Noticed that we have `OmitOverUnion` which works with union types. Moved `OmitOverUnion` to `Utils.ts` file and using it instead.

File: packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts [hunk 2]
```diff
@@ -141,47 +141,59 @@ export type StandardHierarchyNodeKey = InstancesNodeKey | GroupingNodeKey;
...
     return !!(key as StandardHierarchyNodeKey).type;
   }
   /** Checks whether the given node key is an [[InstancesNodeKey]]. */
-  export function isInstances(key: HierarchyNodeKey): key is InstancesNodeKey {
+  export function isInstances(key: ParentGroupingHierarchyNodeKey | HierarchyNodeKey): key is InstancesNodeKey {
     return isStandard(key) && key.type === "instances";
   }
   /** Checks whether the given node key is a [[GroupingNodeKey]]. */
-  export function isGrouping(key: HierarchyNodeKey): key is GroupingNodeKey {
+  export function isGrouping(key: ParentGroupingHierarchyNodeKey | HierarchyNodeKey): key is GroupingNodeKey {
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts]: Moved `compare` and `equals` functions to `ParentNodeKey` namespace.  And removed the `ParentGroupingHierarchyNodeKey | HierarchyNodeKey`.

File: packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts [hunk 3]
```diff
@@ -319,7 +331,7 @@ export interface HierarchyNode {
   /** An identifier to identify the node in its hierarchy level. */
   key: HierarchyNodeKey;
   /** Identifiers of all node ancestors. Can be used to identify a node in the hierarchy. */
-  parentKeys: HierarchyNodeKey[];
+  parentKeys: Array<ParentGroupingHierarchyNodeKey | InstancesNodeKey | string>;
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts]: Changed.

File: packages/hierarchy-builder/src/test/internal/operators/grouping/BaseClassGrouping.test.ts [hunk 4]
```diff
@@ -129,20 +129,22 @@ describe("BaseClassGrouping", () => {
...
         grouped: [
           createTestProcessedGroupingNode({
             label: eCClass.name,
-            key: expectedGroupingNodeKey,
+            key: {
+              ...expectedGroupingNodeKey,
+              groupedInstanceKeys: nodes.flatMap((n) => n.key.instanceKeys),
+            },
             parentKeys: ["x"],
             children: nodes.map((n) => ({ ...n, parentKeys: [...n.parentKeys, expectedGroupingNodeKey] })),
```
[the remark this answers is not in the document]
Comment (JonasDov) [reply packages/hierarchy-builder/src/test/internal/operators/grouping/BaseClassGrouping.test.ts]: Changed to use `omit` where possible.

File: packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts [hunk 5]
```diff
@@ -311,6 +283,47 @@ export namespace HierarchyNodeKey {
   }
 }
 
+/** @beta */
+// eslint-disable-next-line @typescript-eslint/no-redeclare
+export namespace HierarchyNodeKey {
```
Comment (grigasp) [inline packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts:288]: please move this right under the `HierarchyNodeKey` type

File: packages/hierarchy-builder/src/hierarchy-builder/Utils.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+export type OmitOverUnion<T, K extends PropertyKey> = T extends T ? Omit<T, K> : never;
```
Comment (grigasp) [inline packages/hierarchy-builder/src/hierarchy-builder/Utils.ts:6]: needs a release tag

File: packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts (hunk 5 above)
Comment (JonasDov) [reply packages/hierarchy-builder/src/hierarchy-builder/HierarchyNode.ts]: Moved.

File: packages/hierarchy-builder/src/hierarchy-builder/Utils.ts (hunk 6 above)
Comment (JonasDov) [reply packages/hierarchy-builder/src/hierarchy-builder/Utils.ts:7]: Added internal tag.

Comment (grigasp) [reply packages/hierarchy-builder/src/hierarchy-builder/Utils.ts:7]: isn't it exposed by `@beta`?

Comment (JonasDov) [reply packages/hierarchy-builder/src/hierarchy-builder/Utils.ts:10]: Changed to `@beta`

**Review by grigasp — APPROVED** [review]

**Pull request merged**